scholarly/_navigator.py

```diff
@@ -0,0 +1,195 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from bs4 import BeautifulSoup
+
+import codecs
+import hashlib
+import logging
+import random
+import re
+import requests
+import time
+from stem import Signal
+from stem.control import Controller
+from fake_useragent import UserAgent
+from .publication import _SearchScholarIterator
+from .author import Author
+import sys
+
+_GOOGLEID = hashlib.md5(str(random.random()).encode('utf-8')).hexdigest()[:16]
+_COOKIES = {'GSP': 'ID={0}:CF=4'.format(_GOOGLEID)}
+_HEADERS = {
+    'accept-language': 'en-US,en',
+    'accept': 'text/html,application/xhtml+xml,application/xml'
+}
+_HOST = 'https://scholar.google.com{0}'
+
+_SCHOLARCITERE = r'gs_ocit\(event,\'([\w-]*)\''
+
+_TIMEOUT = 2
+
+_PROXIES = {
+    "http": None,
+    "https": None,
+}
+
+_HTTP_PROXY = None
+_HTTPS_PROXY = None
+
+_TOR_SOCK = None
+_TOR_CONTROL = None
+
+
+if sys.platform.startswith("linux"):
+    _TOR_SOCK = "socks5://127.0.0.1:9050"
+    _TOR_CONTROL = 9051
+elif sys.platform.startswith("win"):
+    _TOR_SOCK = "socks5://127.0.0.1:9150"
+    _TOR_CONTROL = 9151
+
+class Navigator(object):
+    """I did not call it browser because there are other packages with
+    that exact name. -Victor
+    """
+    def __init__(self):
+        #TODO: Implement Singleton pattern since we don't need multiple navs.
+        super(Navigator, self).__init__()
+        logging.basicConfig(filename='scholar.log', level=logging.INFO)
+        self.logger = logging.getLogger('scholarly')
+
+
+    def _get_page(self, pagerequest:str):
+        """Return the data for a page on scholar.google.com"""
+        self.logger.info("Getting %s", pagerequest)
+        resp = None
+        while True:
+            # If Tor is running we use the proxy
+            # Did not use with for shorter indented lines -V
+            session = requests.Session()
+            if self._tor_works():
```

Comment:
You may want to store the output of `_tor_works()` in a variable instead of testing if tor works in every call (and therefore issuing an additional network request to fetch Google)